---
title: "Class15: Pathway Analysis from RNA-Seq Results"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## RNA-Seq analysis 

The data for for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:

> Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703


```{r}
library(DESeq2)
```

```{r}
metaFile <- "GSE37704_metadata.csv"
countFile <- "GSE37704_featurecounts.csv"

# Import metadata and take a peak
colData = read.csv(metaFile, row.names=1)
head(colData)
```

```{r}
# Import countdata
countData = read.csv(countFile, row.names=1)
head(countData)
```

Hmm... looks like we need to remove the first "length" column in the `countData` object.

```{r}
# remove first col from countData
countData <- as.matrix(countData[,-1])
head(countData)
```

Let's remove the rows with zero counts in all experiments (i.e. columns)

```{r}
nonzero.rows <- rowSums(countData) != 0 
countData <- countData[nonzero.rows,]
```


```{r}
nrow(countData)
```

## Running DESeq2
Nice now lets setup the DESeqDataSet object required for the DESeq() function and then run the DESeq pipeline. This is again similar to our last days hands-on session.

```{r}
dds = DESeqDataSetFromMatrix(countData=countData,
                             colData=colData,
                             design=~condition)
dds = DESeq(dds)
```

Get our results

```{r}
res <- results(dds)
summary(res)
```

## Volcono plot
Now we will make a volcano plot, a commonly produced visualization from this type of data that we introduced last day. Basically it's a plot of log2 fold change vs -log adjusted p-value.

```{r}
plot(res$log2FoldChange, -log(res$padj))
abline(v=c(-2,2), col="green", lty=2, lwd=2)
abline(h=-log(0.01), col="green", lty=2, lwd=2)
```

And with color points


```{r}
# Define a vector of colors
mycols <- rep("lightgray", nrow(res))
inds <- (abs(res$log2FoldChange) > 2) & res$padj < 0.01
mycols[ inds ] <- "blue"

plot(res$log2FoldChange, -log(res$padj), col=mycols)
abline(v=c(-2,2), col="black", lty=2, lwd=2)
abline(h=-log(0.01), col="black", lty=2, lwd=2)
```

### Add annotation (i.e. different gene names and ids)
Use the mapIDs() function multiple times to add SYMBOL, ENTREZID and GENENAME annotation to our results by completing the code below

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

columns(org.Hs.eg.db)
```

```{r}
res$symbol = mapIds(org.Hs.eg.db,
                    keys=row.names(res), 
                    keytype="ENSEMBL",
                    column="SYMBOL",
                    multiVals="first")

res$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")

res$name =   mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="GENENAME",
                    multiVals="first")

```

```{r}
head(as.data.frame(res))
```

Write out our ordered and annotated results object

```{r}
res = res[order(res$pvalue),]
write.csv(res, file="deseq_results.csv")
```

Time estimate 11am



## Section 2. Pathway analysis

```{r}
# Install packages in your CONSOLE!
# BiocManager::install( c("pathview", "gage", "gageData") )
```

```{r}
library(pathview)
library(gage)
library(gageData)
```

```{r}
data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

```

```{r}
head(kegg.sets.hs, 3)
```

Create a vector `foldchanges` of FoldChange values that has ENTREZ identifiers as the names of the vector. This is the format that the **gage()** function wants.

```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

Run the **gage()** function with this input

```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

Now lets look at the object returned from gage().
```{r}
attributes(keggres)
```

```{r}
head(keggres$less)
```


```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```

![](hsa04110.pathview.png)


```{r}
## Focus on top 5 upregulated pathways here for demo purposes only
keggrespathways <- rownames(keggres$greater)[1:5]

# Extract the 8 character long IDs part of each string
keggresids = substr(keggrespathways, start=1, stop=8)
keggresids
```


```{r warning=FALSE, message=FALSE}
pathview(gene.data=foldchanges, pathway.id=keggresids, species="hsa")
```

Here are the plots:

![](hsa00140.pathview.png)

![](hsa04142.pathview.png)

![](hsa04630.pathview.png)

![](hsa04640.pathview.png)



## GO analysis

We can also do a similar procedure with gene ontology. Similar to above, go.sets.hs has all GO terms. go.subs.hs is a named list containing indexes for the BP, CC, and MF ontologies. Letâ€™s focus on BP (a.k.a Biological Process) here.

```{r}
data(go.sets.hs)
data(go.subs.hs)

# Focus on Biological Process subset of GO
gobpsets = go.sets.hs[go.subs.hs$BP]

gobpres = gage(foldchanges, gsets=gobpsets, same.dir=TRUE)

lapply(gobpres, head)

```

## Section 4. Reactome Analysis

Reactome is database consisting of biological molecules and their relation to pathways and processes. Reactome, such as many other tools, has an online software available (https://reactome.org/) and R package available (https://bioconductor.org/packages/release/bioc/html/ReactomePA.html).

First, Using R, output the list of significant genes at the 0.05 level as a plain text file:

```{r}
sig_genes <- res[res$padj <= 0.05 & !is.na(res$padj), "symbol"]
print(paste("Total number of significant genes:", length(sig_genes)))
```

```{r}
write.table(sig_genes, file="significant_genes.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)
```



