---
title: "Class 13: Genome Informatics (Pt 1.)"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Sample genotypes in the MXL 1000 Genome project data

Here we focus on the Mexican Ancestry in Los Angeles, California (MXL) population

What proportion of the Mexican Ancestry in Los Angeles sample population (MXL) are homozygous for the asthma associated SNP (G|G) rs8067378 

```{r}
# Read CSV from ENSEMBLE
mxl <- read.csv("373531-SampleGenotypes-Homo_sapiens_Variation_Sample_rs8067378.csv")
head(mxl)

```

How many of each genotype are there?
```{r}
table(mxl$Genotype..forward.strand.)
```

Proportion or percent of total for each genotype

```{r}
(table(mxl$Genotype..forward.strand.) / nrow(mxl) ) * 100
```


## Quality Scores in FASTQ files

The forth line of a FASTQ sequence format file encodes the quality score that tells us how good the sequence at a give position is (i.e. how likely it is to be correct based on the instrument)


```{r}
library(seqinr)
library(gtools)

#s2c("DDDDCDEDCDDDDBBDDDCC@")
asc( s2c("DDDDCDEDCDDDDBBDDDCC@") ) - 33
#phred

```


## Population Scale analysis

Read RNA-Seq count data with genotype information results table

```{r}
expr <- read.table("rs8067378_ENSG00000172057.6.txt")

head(expr)
```


```{r}
summary(expr)
```


```{r}
inds <- expr$geno == "G/G"
summary(expr[inds, "exp"])
```

```{r}
inds <- expr$geno == "A/G"
summary(expr[inds, "exp"])

```

```{r}
inds <- expr$geno == "A/A"
summary(expr[inds, "exp"])

```


```{r}
boxplot(exp ~ geno , data=expr)
```


```{r}
# Generate a boxplot for expression data of different genotype groups
p <- boxplot(exp~geno, data=expr, xlab="rs8067378 genotype", ylab="ENSG00000172057.4 (RPKM)", notch=T)
```


And maybe ggplot

```{r}
library(ggplot2)

## Histogram of the exp column with ggplot2
ggplot(expr, aes(exp, fill = geno)) + geom_density(alpha = 0.2)
```

```{r}
# Boxplot with the data shown
ggplot(expr, aes(geno, exp, fill=geno)) + 
  geom_boxplot(notch=TRUE, outlier.shape = NA) + 
  geom_jitter(shape=16, position=position_jitter(0.2), alpha=0.4)
```

